로그인 컴포넌트 만들기

xunxee opened the pull request:
# 🔴 구현 사항 설명
## 🟠 10월 8일
- Webpack, Babel 을 이용하여 초기 세팅을 진행
## 🟠 10월 9일
- "react-redux" 를 사용하여 상태 관리
- "redux toolkit" 의 createSlice 를 사용하여 slice.js 구현
- LoginFormContainer 컴포넌트에서 isLogin 의 값에 따라 조건부 렌더링이 일어나도록 컴포넌트를 구현
## 🟠 10월 10일
- initial state 의 state 의 이름을 가독성 있게 변경(isLoginState -> isLogin)
- useCallback 으로 최적화된 함수를 받는 컴포넌트를 "memo" 를 사용하여 최적화 진행
- login 과 signup 을 관리하는 컴포넌트를 하나로 관리하여 불필요한 렌더링을 관리
- 테스트 코드에서 given 을 사용하여, 테스트에 사용하는 변수를 정의
## 🟠 10월 11일
- 10월 10일의 리뷰 내용을 반영
## 🟠 10월 12일
- 10월 11일의 리뷰 내용을 반영
- configureStore 를 사용하여 reducer 를 분리
- LoginForm 컴포넌트의 change 이벤트를 구현

xunxee committed:
Modify: github에 올라간 node_modules 디렉터리를 삭제

xunxee committed:
Refactor: 리뷰 내용을 반영하여 리팩터링 완료
- github 의 "No newline at end of file" 에러를 해결
- package.json 의 dependencies 에 사용하지 않는 것들을 정리
- index.html 의 title 을 ShopifyStore 로 변경

xunxee committed:
Add: loginForm 의 기본적인 형태를 렌더링하는 액션을 테스트

xunxee committed:
Add: 기본적인 Route 를 구현하는 액션을 테스트
- HomePage, LoginPage, NotFoundPage 를 추가
- MemoryRouter 를 사용하여 액션을 테스트

xunxee committed:
Add: HomePage 컴포넌트에  Link 로 이동하는 액션을 테스트

xunxee committed:
Add: LoginPage 의 조건부 렌더링을 위한 액션을 테스트
- 상태 관리를 위해 "react-redux" 를 사용
- redux toolkit 의 creacteSlice 를 사용하여 slice.js 를 구현
- LoginFormContainer 컴포넌트에서 isLoginState 의 value 에 따라 조건부 렌더링이 일어나도록 구현

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -0,0 +1,29 @@
+    const { container } = renderLoginForm();
+
+    expect(container).toHaveTextContent('have an account?');
+  });
```
describe -> it 형식은 말로 쭉 이어서 써봤을 때 말이 되게 하면 좋다고 하더라구요!
describe는 보통 주어가 되고 it은 describe가 무엇을 하는지 적어주면 됩니당!!

적어 놓은 신대로 풀어서 적어보면 LoginForm renders the LoginForm. 이렇게 돼서 LoginForm이 LogrinForm을 렌더링한다. 라고 됩니다!~ 뭔가 어색하죠??

이런 겨우엔 LoginForm renders the login fileds. 라고 작성해주셔도 좋을 거 같아요! 다희님은 무조건 영어로 테스트 코드 작성을 권장하셨지만 이동욱님도 보면 한글로 작성해주셨었거든요! 그래서 편하신대로 해주시면 될 거 같아요!!

그리고 LoginForm은 제일 하단 component라서 테스트를 좀 더 구체적으로 작성해주시면 좋을 거 같습니당!ㅎㅎ
최대한 모든 요소를 다 확인해준다는 느낌으로요!

xunxee replied:
describe 를 주어로 생각하니깐 훨씬 와닿는 것 같네요!

LoginForm 컴포넌트의 관심사가 "렌더링" 이다보니 화면에 그려지는 모양들을 더 살펴볼 수 있도록 테스트를 추가했습니다 🐥
감사합니닷 🙏🏼

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -0,0 +1,29 @@
+    expect(container).toHaveTextContent('have an account?');
+  });
+
+  it('clicks Sing up button', () => {
```
이 부분도 말씀드린 것과 비슷하게 생각하시면 좋을 거 같아요!
LoginForm clicks sign up button. -> 'LoginForm이 버튼을 클린한다' 어색하죠?! 영어로 작성하는 경우엔 'renders sign up button to listen click event' or 'listens click event' 이런식으로 많이 작성 하더라라구요!

xunxee replied:
생각해보니 한글로 만약 작성했다면 자연스럽게 말씀하신 것 처럼 작성을 했을 것 같은데
영어로 쓰다보니 인지를 못했던 부분이네요 ㅠㅠ
감사합니당!!

ChoJinmok replied:
영어로 끝까지 한번 가보시죠~!ㅎㅎ

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+import { useCallback } from 'react';
+
+import { useDispatch, useSelector } from 'react-redux';
+import LoginForm from './LoginForm';
```
한 줄 띄워주면 좋을 거 같습니당!!ㅎㅎ

xunxee replied:
앗 네넵!!

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+import LoginForm from './LoginForm';
+import SignUpForm from './SignUpForm';
+
+import { setIsLoginState } from '../slice';
```
state이름을 isLogin으로만 해줘도 state가 들어나니까 state를 빼줘도 좋을 거 같아요!!🙂

xunxee replied:
고민을 하긴했었는데, 그 자체로 state 를 만드는 것이니깐 생략하는 것이 맞는 것 같네용!! 😆

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+
+  const isLoginState = useSelector((selector) => selector.isLoginState);
+
+  console.log(isLoginState);
```
push 하시기 전에 'npx eslint --fix .' 해주시면 불필요한 요소들도 잡아줘서 확인해주면 좋습니당!! 동작한다면요.. 😭

xunxee replied:
ㅜㅜ 하핫 왜 동작을 안하는지 모르겠네요 🥺
push 전에 수동으로 확인을 해야겟습니당!

ChoJinmok replied:
언제 만나서 같이 한번 해보시죠 ㅠㅠ

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -0,0 +1,25 @@
+export default function LoginForm({ onClick }) {
```
useCallback으로 최적화된 함수를 받는 경우 그 컴포넌트도 최적화 진행해주면 좋습니당!!ㅎㅎ

xunxee replied:
그렇군요! useCallback 을 받는 컴포넌트들을 memo 사용하여 최적화 진행했습니다!

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -0,0 +1,29 @@
+        <LoginForm onClick={handleClickLoginState} />
+      ) : (
+        <SignUpForm onClick={handleClickLoginState} />
+      )}
```
두개의 컴포넌트로 아예 나누는 것보다는 isLogin을 props로 넘겨준다음에 하나의 컴포넌트로 관리해주는게 좋을 거 같아요!!

xunxee replied:
두 개를 컴포넌트를 각각 모두 렌더링하는 것이 아니라,
하나의 컴포넌트 안에서 조건부 렌더링을 사용하여 필요한 부분만 추가적으로 렌더링할 수 있도록 해야겠군용 🙏🏼

ChoJinmok commented on src/LoginPage/LoginFormContainer.test.jsx:
```diff
@@ -0,0 +1,41 @@
+jest.mock('react-redux');
+
+describe('LoginFormContainer', () => {
+  context('When the value of isLoginState is true', () => {
```
'with logged in' 이런식으로 더 간단하게 표현해줄 수 있을 거 같아요!ㅎㅎ

ChoJinmok commented on src/LoginPage/LoginFormContainer.test.jsx:
```diff
@@ -0,0 +1,41 @@
+      useSelector.mockImplementation((selector) => selector({
+        isLoginState: true,
+      }));
+    });
```
given 설치 하셨으니까 한번 사용해보시면 좋겠네요 🙂

xunxee replied:
오... given 좋은 것 같군용 😊 감사합니당

ChoJinmok replied:
리뷰 반영하신 거 봤는데 아주 절절하게 사용하셨던데요?!👍

ChoJinmok commented on src/LoginPage/SignUpForm.jsx:
```diff
@@ -0,0 +1,33 @@
+        name="last"
+        type="text"
+        placeholder="Last Name"
+      />
```
```suggestion
{isLogin && (
<input
        name="first"
        type="text"
        placeholder="First Name"
      />
      <input
        name="last"
        type="text"
        placeholder="Last Name"
      />
)}
```

ChoJinmok replied:
만약 하나의 컴포넌트로 작성해주시면 이런식으로 해줄 수 있을 거 같아요!~

xunxee replied:
넵 조건부 렌더링 사용해서 하나의 컴포넌트로 해야겠습니다!

ChoJinmok replied:
리뷰 보고 한번 확인해주세용! ㅎㅎ

ChoJinmok reviewed (commented):
아주 잘하고 계신데요?~ ㅎㅎ TDD지키시면서 끝까지 팟팅입니당!🙂
Container 컴포넌트에서도 click event 테스트 해보시면 좋을 거 같네용!~ㅎㅎ

xunxee committed:
Refactor: 10월 9일 리뷰를 반영
- describe -> it 으로 이어지는 형식의 표현을 수정
- import 하는 구문을 관심사 별로 나누는 작업을 진행
- initial state 의 state 의 이름을 가독성 있게 변경(isLoginState -> isLogin)
- useCallback 으로 최적화된 함수를 받는 컴포넌트를 "memo" 를 이용하여 최적화 진행
- login 과 signup 을 관리하는 컴포넌트를 하나로 관리하여 모든 컴포넌트가 렌더링되는 것을 리팩터링
- 테스트 코드에서 given 을 사용하여, 테스트에 사용하는 변수를 정의

xunxee committed:
slice 의 setIsLogin 액션을 테스트

xunxee committed:
Add: combineReducers 를 사용하여 slice 를 나누는 작업을 시작

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -1,46 +1,55 @@
-      : (<button type="submit">Sign Up</button>)}
+export default memo(({ isLogin, onClick, onChange }) => {
+  function handleChange(event) {
+    const { target: { name, value } } = event;
```
```suggestion
function handleChange({ target: { name, value }}) {
```
알고 계실 거 같긴한데~ 저같은 경우엔 이렇게 바로 적어주는게 더 편하긴 하더라구요~ㅎㅎ 이건 참고사항입니다!

xunxee replied:
엇? 저는 생각 못한 방법인데요??? 감사합니다🙏🏼

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -1,46 +1,55 @@
-)));
+  return (
+    <form>
+      {isLogin ? null : (
```
```suggestion
{isLogin && (
```
이런식으로 단축평가 사용하면 동작을 하지 않던가용?~🤔

xunxee replied:
login 과 signup 버튼을 누를 때마다 text 가 바뀌어야 하는 부분들을 구현하다보니 자연스럽게 그 부분도 삼항 연산자를 사용했는데
알려주신 부분에서는 "논리곱 연산자" 를 사용하는 것이 맞겠네요! 😆

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -1,46 +1,55 @@
+      />
+      {isLogin ? (
+        <button type="submit">Log In</button>)
+        : (<button type="submit">Sign Up</button>)}
```
```suggestion
<button type="submit">{isLogin ? 'Log In' : 'Sign Up'}</button>
```

아래쪽 버튼처럼 작성해주시는게 렌더링, 가독성 측면에서 더 좋을 거 같아요!
아래 p태그도 한번 고쳐주세용!ㅎㅎ

xunxee replied:
이 부분들을 저렇게 고쳤어야 했는데 몰랐네요
덕분에 코드가 훨씬 가독성이 좋아졌습니다!
🙏🏼

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -31,6 +33,20 @@ describe('LoginForm', () => {
+      });
+
+      const input = getByPlaceholderText('Email');
+      const value = { value: 'test@test.com' };
```
```suggestion
const value = 'test@test.com';
```
그냥 이렇게 작성하셔도 똑같이 동작할 거 같아요!!🙂

xunxee replied:
아앗 그렇네요!
테스트 코드의
```jsx
expect(handleChange).toBeCalledWith({ name, value });
```
에 맞게 해주려다가 RORO 패턴과 혼동햇네요! 🥹

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -15,11 +15,16 @@ export default function LoginFormContainer() {
 
+  const handleChange = useCallback(({ name, value }) => {
+    dispatch(changeLoginFileds());
+  }, [name, value]);
```
name, value는 handleChange의 매개변수여서 name, value를 외부에 의존한다고는 할 수 없을 거 같아요!
코드상에서 외부에 의존하는 건 dispatch, changeLoginFields 두가지 인데 changeLoginFields의 경우엔 저희가 직접 작성해서 절대 변할 일이 없다고 확신할 수 있지만, dispatch의 경우엔 redux에서 제공해주는 함수라 변할 지 변하지 않을 지는 저희가 장담할 수 없기 때문에 dispatch만 의존성 배열에 넣어줘야합니다!
결국엔 함수의 코드상에 변할 가능성이 있는 요소만 의존성 배열에 넣어주시면 돼요! 정~말 그럴일이 없겠지만 useDisptch가 특정 경우엔 다른 dispatch를 반환할 수도 있기 때문이죠! 코드를 작성한 사람도 햇갈리는데 외부에서 작성한 코드를 100% 믿을 수 없는 문젠 거 같기도 합니다! 결국 다른 dispatch가 들어오는 순간 useCallback은 함수를 새롭게 만들어야하는 상황인거죠! 만약 dispatch가 변하면 함수의 결과가 달라지겠지만 useCallback의 의존성배열에 dispatch가 들어가 있지 않으면 함수를 새롭게 생성하지 않을 거라서 원하는대로 동작을 하지 않게됩니당!

ps. changeLoginFileds 오타 발견했네용 ㅎㅎ

xunxee replied:
useCallback 제가 잘못 이해하고 있는 부분이 있었군요
감사합니닷!! 🥺

ChoJinmok commented on src/reducer.js:
```diff
@@ -0,0 +1 @@
+import { combineReducers } from 'redux';
```
오늘 이야기 했던 부분이군요!?

xunxee replied:
맞습니당 ㅎㅎ
configureStore() 를 사용해서 slice 들을 나누고 모아주는 방향으로 해보겠습니다!

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -4,15 +4,15 @@ import { useDispatch, useSelector } from 'react-redux';
 export default function LoginFormContainer() {
   const dispatch = useDispatch();
 
   const isLogin = useSelector((selector) => selector.isLogin);
```
```suggestion
const isLogin = useSelector(({ isLogin }) => isLogin);
```
이것도 취향차이긴 한데 이렇게 작성하면 조금 더 간결해지긴 합니당!ㅎㅎ

xunxee replied:
아? selector 가 slice 의 inintialState 객체기 때문에 객체 구조분해할당이 가능하군용
이 부분 생각하지 못했네요!
디스트럭처링해서 selector 도 사용해야겠습니다! 🙏🏼

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -15,11 +15,16 @@ export default function LoginFormContainer() {
     dispatch(setIsLogin());
   }, []);
```
이부분도 의존성배열에대해서 길게 적어놓은 부분 보시고 여기도 한번 고쳐보시면 좋을 거 같아요!

ChoJinmok reviewed (commented):
pr 일자별로 작성하신 거 너무 깔끔합니당!! 저도 적용해봐야겠네요..ㅎ
테스트 코드 작성 잘하고 계신데요!?!?! LoginForm test 아주 좋았던 거 같아요!!ㅎㅎ 오늘도 고생하셨습니다!~👍

xunxee committed:
Refactor: 리뷰 내용을 반영

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -4,20 +4,20 @@ import { useDispatch, useSelector } from 'react-redux';
-  const isLogin = useSelector((selector) => selector.isLogin);
+  const { isLogin } = useSelector((selector) => selector);
 
   const handleClickLoginState = useCallback(() => {
```
toggle button을 클릭해주는 함수는 함수명에 toggle이 자주 사용되더라구요!
함수명 handleClickLoginState만 보면 LoginState를 변화시켜주는 느낌보다는 LoginState를 클릭하는 느낌이 많이 드는 거 같아요.
handleClickToggle 이나 handleToggleLoginState가 조금 더 자연스러울 거 같아요! ㅎㅎ 참고하세용~

xunxee replied:
그러네요 함수명 handleClickLoginState 이 뭔가 명확하지 않은 느낌이 확실히 있네요
toggle 을 이용해서 함수명을 수정해보겠습니다!
감사합니당 ㅎㅎ

ChoJinmok commented on src/store.js:
```diff
@@ -1,7 +1,9 @@
-const store = configureStore({ reducer });
+const store = configureStore({
+  reducer: rootReducer,
+});
```
```javascript
import loginReducer from './LoginPage/slice';

const store = configureStore({
  reducer: {
    login: loginSlice
    // 이후에 다른 slice들 추가
  },
});
```

이런식으로 작성하는 방법 한번 보시면 좋을 거 같네용!~

<br />
rootReducer를 사용해주는 경우는

```javascript
import loginReducer from './LoginPage/slice';

const rootReducer = combineReducers({
  loginReducer,
  // 이후에 다른 slice들 추가
});

const store = configureStore({
  reducer: rootReducer,
});
```
이럴 때일 거 같습니당! ㅎㅎ 그런데 configureStore는 1단계 깊이의 객체의 경우 자동으로 combineReducers로 묶어주기 때문에 더 복잡해지는 경우라면 combineReducers가 필요하겠지만 아직은  combineReducers가 필요하진 않을 거 같아요!~

<br />
어찌됐든 두가지 방법 모두 useSellector로 state를 받아오는 방식이 달라져서 state가져오는 연습이 되실 거 같아요!🙂

ChoJinmok reviewed (commented):
store에 configureStore에 reducer 넣어주는 방식 한번 더 확인해보시면 좋을 거 같아요!! 팟팅입니당 👍

xunxee committed:
Refactor: 10월 11일의 리뷰를 반영

xunxee committed:
Add: configureStore 를 사용하여 reducer 를 나누는 액션을 테스트
- 기존의 통합적으로 관리하던 slice 를 configureStore 를 사용하여 분리하는 작업을 진행했습니다.

xunxee committed:
Add: LoginFrom 컴포넌트의 change 이벤트를 구현
- "Log In" 과 "Sign Up" 각각의 상황에서 사용자가 입력한 값을 저장할 수 있도록 change 이벤트를 구현했습니다.

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -1,6 +1,15 @@
 import { memo } from 'react';
 
-export default memo(({ isLogin, onClick, onChange }) => {
+export default memo(({
+  isLogin,
+  fields,
+  onClick,
+  onChange,
+}) => {
+  const {
+    email, password, firstName, lastName,
+  } = fields;
```
```suggestion
export default memo(({
  isLogin,
  fields: { email, password, firstName, lastName },
  onClick,
  onChange,
}) => {
```
이런식으로 작성해도 편하더라구요 ㅎㅎ 참고하세용!~

xunxee replied:
앗 그렇군요! 어차피 아래에서 한번더 구조분해할당을 사용해야하는데
바로 사용할 수 있겠군요!
생각하지 못한 부분입니다! 감사합니당 🙏🏼

ChoJinmok commented on src/LoginPage/LoginFormContainer.jsx:
```diff
@@ -4,25 +4,37 @@ import { useDispatch, useSelector } from 'react-redux';
   const dispatch = useDispatch();
 
   const { isLogin } = useSelector(({ login }) => login);
+  const {
+    email,
+    password,
+    firstName,
+    lastName,
+  } = useSelector(({ login: { loginFields } }) => loginFields);
```
```javascript
const loginFields = useSelector(({ login: { loginFields } }) => loginFields);
```
이런식으로 작성하고 밑에 LoginForm에 fields props에 loginFields 그대로 넣어주셔도 될 거 같아요 ㅎㅎ

xunxee replied:
utils.js 파일을 하나 만들어서 저 부분을 함수화해서 해결해보아야 겠습니다!

ChoJinmok commented on src/LoginPage/slice.test.js:
```diff
@@ -17,4 +18,29 @@ describe('reducer', () => {
+          },
+        };
+
+        const state = reducer(
```
```javacript
const { loginFields: { email } } = reducer(
```

저는 이렇게 해서 조금더 편하더라구요!? 한번 참고하세용~ㅎㅎ

xunxee replied:
아 그렇군요! 
이 방법 사용하는게 정말 많이 활용되는군용 ㅠㅠ
생각하지 못했네요 🥹
감사합니당! 가독성이 더 높아질 것 같아요!

ChoJinmok reviewed (commented):
코드적으로 크게 리뷰드릴게 없는데요!?ㅎㅎ이제 login page는 마무리 단계인 거 같네용 팟팅입니다!~ㅎㅎ
더 시보해보셨으면 좋겠는게
1. slice test에서 reducer에 초기 state가 설정되는 테스트도 추가해보시면 좋을 거 같아요!(코드숨 참고)
2. App test에 경로 테스트 더 추가해보는 것도 좋을 거 같습니당!!ㅎㅎ
    - login, not found page 추가
